src/djinterop/engine/v2/beat_data_blob.cpp

```diff
@@ -0,0 +1,176 @@
+/*
+    This file is part of libdjinterop.
+
+    libdjinterop is free software: you can redistribute it and/or modify
+    it under the terms of the GNU Lesser General Public License as published by
+    the Free Software Foundation, either version 3 of the License, or
+    (at your option) any later version.
+
+    libdjinterop is distributed in the hope that it will be useful,
+    but WITHOUT ANY WARRANTY; without even the implied warranty of
+    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+    GNU Lesser General Public License for more details.
+
+    You should have received a copy of the GNU Lesser General Public License
+    along with libdjinterop.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+#include <djinterop/engine/v2/beat_data_blob.hpp>
+
+#include <cassert>
+
+#include "../encode_decode_utils.hpp"
+
+namespace djinterop::engine::v2
+{
+namespace
+{
+char* encode_beatgrid(
+    const std::vector<beat_grid_marker_blob>& beatgrid, char* ptr)
+{
+    typedef std::vector<beat_grid_marker_blob>::size_type vec_size_t;
```

Comment:
Encoding/decoding code is repeated here, adapted from the v1 codebase.  It is slightly different, however, as the code here works with structs that _directly_ map to the encoded format.  The v1 equivalent code had to both encode/decode _and_ normalise at the same time, which is a conflation of concerns.

It is expected that the v1 codebase ought to be refactored to this approach at some point, and then perhaps these implementations can be moved out of the `engine::v2` namespace and into the shared `engine` namespace instead.